### ISSUE TITLE
Atomic layout updates (eliminate flicker on window resize)

### ISSUE THREAD

**Issue by djpohly** [djpohly/dwl#19]
When the window layout changes, dwl should not render a frame with the new layout until all of the visible windows can be drawn at their new sizes.

Currently it is possible to draw frames in which the size/position/visibility of windows are in an intermediate state.  For example, when a new tiled window is mapped, the other tiled windows may commit their new-sized buffer and be redrawn in their updated positions before the new window appears, showing the background through a "hole," or a window which needs to be sized down will be drawn in its new position before its buffer changes size, overlapping with other windows.

The solution doesn't seem to be short and obvious, so if there is not a simple and fully correct way to do this, then let's aim for simple and sufficient before complex and fully correct.

**Comment by alex-courtis** [comment]
A quick debug has confirmed that the same thread is invoking rendermon as invokes tile. Therefore we are not looking at a simple problem of concurrent data access.

Confirming my understanding of the example above: A new surface is created, and is not ready to be rendered at createnotify time, but is ready once maprequest is invoked.

Solutions for that example might include:
1. Block rednermon between createnotify and maprequest
2. Do not include the new surface in the client list until maprequest
3. Block arrange functions between createnotify and maprequest

The general solution involves identifing pairs of signal functions that demarcate these intermeditate states and avoiding display of those states.

Please advise whether this is a correct summary of the problem, or if there is something bigger that I am not seeing.

**Comment by djpohly** [comment]
Correct, this isn't about concurrent data access.  Since Wayland is asynchronous, there's time (and potentially frame-render requests) during the "configure - ack_configure - commit" cycle for each window that is being created or resized, and that time may differ for each window.  `arrange` sends the initial configure to all of the windows, but we don't want to draw *any* window at its new size/position until *all* of the windows have committed at their new sizes (or, at least, until we can draw something reasonable at their new sizes).

**Comment by alex-courtis** [comment]
Understood.

Sway has a transaction system that blocks render during transactions, although that appears to be more to handle sway's arrangement of windows being spread across many different places.

Transactions are a very heavy solution; something simpler such as dirty client state might suffice for dwl.

**Comment by djpohly** [comment]
I suspect we'll have to add a surface.commit handler and do something with `configure_serial`.  We don't have a buffer at the new size until the first commit after ack_configure, and it doesn't look like wlroots does anything special at commit time to record which xdg_shell state changes have actually been acknowledged.  swaywm/wlroots#762 and swaywm/wlroots#1657 both sound related.

**Comment by alex-courtis** [comment]
A surface commit handler can test:

```c->dirty = c->xdg_surface->configure_serial != c->xdg_surface->configure_next_serial;```

Logging indicates that the new size is ready when the commit handler is called with the above two equal. They are equal at other times e.g. rendermon, however that doesn't indicate readiness.

Problem now: how to prevent rendering when there is a dirty client. All my attempts at interrupting rendermon have failed so far as rendermon stops being called...

**Comment by djpohly** [comment]
#28 seems to address the flicker for resizes, but appears not to completely fix it for creating/destroying windows.  I ran

    for n in `seq 20`; do alacritty -e sleep 0.02 & done

from an alacritty terminal running in that dwl session (so there was always at least one window open), but I still saw the background color flicker in.  I haven't determined whether it was on create or destroy or both.

**Comment by alex-courtis** [comment]
Interesting... I can see the background colour flashes when I run the above with st (as I expected) but not alacritty. Perhaps a cpu vs gpu race.

I'll instrument some more and see if there's a configure event on startup / shutdown that we can look for.

**Comment by alex-courtis** [comment]
I've added some debug, to try and see what's going on with the commits.

@djpohly please run the branch and manually create three alacritty instances in tiled mode.

Here's what I got from running dwl from the console, using intel integrated graphics:
```
[Detaching after fork from child process 37534]
AMC commitnotify 0x55555606d580 c=0 n=1 r=0
AMC resize       0x55555606d580 c=1 n=2 r=2
AMC commitnotify 0x55555606d580 c=1 n=2 r=2
AMC rendermon    0x55555606d580 c=2 n=2 r=2 (skipping)
AMC commitnotify 0x55555606d580 c=2 n=2 r=0
AMC commitnotify 0x55555606d580 c=2 n=2 r=0
AMC commitnotify 0x55555606d580 c=2 n=2 r=0
AMC commitnotify 0x55555606d580 c=2 n=2 r=0
[Detaching after fork from child process 37607]
AMC commitnotify 0x55555606d580 c=2 n=2 r=0
AMC commitnotify 0x55555601a680 c=0 n=17 r=0
AMC resize       0x55555601a680 c=17 n=23 r=23
AMC resize       0x55555606d580 c=2 n=24 r=24
AMC commitnotify 0x55555601a680 c=17 n=23 r=23
AMC commitnotify 0x55555606d580 c=24 n=24 r=0
AMC rendermon    0x55555601a680 c=23 n=23 r=23 (skipping)
AMC commitnotify 0x55555601a680 c=23 n=23 r=0
AMC commitnotify 0x55555606d580 c=24 n=24 r=0
AMC commitnotify 0x55555601a680 c=23 n=23 r=0
AMC commitnotify 0x55555601a680 c=23 n=23 r=0
AMC commitnotify 0x55555601a680 c=23 n=23 r=0
[Detaching after fork from child process 37680]
AMC commitnotify 0x55555601a680 c=23 n=23 r=0
AMC commitnotify 0x55555607a960 c=0 n=36 r=0
AMC resize       0x55555607a960 c=36 n=42 r=42
AMC resize       0x55555601a680 c=23 n=43 r=43
AMC resize       0x55555606d580 c=24 n=44 r=44
AMC commitnotify 0x55555607a960 c=36 n=42 r=42
AMC rendermon    0x55555607a960 c=36 n=42 r=42 (skipping)
AMC rendermon    0x55555601a680 c=23 n=43 r=43 (skipping)
AMC rendermon    0x55555606d580 c=24 n=44 r=44 (skipping)
AMC commitnotify 0x55555606d580 c=44 n=44 r=0
AMC commitnotify 0x55555607a960 c=42 n=42 r=0
AMC commitnotify 0x55555601a680 c=43 n=43 r=0
AMC commitnotify 0x55555601a680 c=43 n=43 r=0
AMC commitnotify 0x55555606d580 c=44 n=44 r=0
AMC commitnotify 0x55555607a960 c=42 n=42 r=0
AMC commitnotify 0x55555607a960 c=42 n=42 r=0
AMC commitnotify 0x55555607a960 c=42 n=42 r=0
```

You can see that sometimes the current and next commits are sometimes equal during render, but we wait until commit to render, as those are still in an indeterminate state.

**Comment by PalanixYT** [comment]
@Sevz17 This has already been fixed

**Comment by Sevz17** [comment]
Fixed in #198
